File: apps/macos/Sources/HyphenTransfer/TransferCheckpointStore.swift
import Foundation

/// Peer-bound durable transfer checkpoints (dimension 02 P2). Stores manifest,
/// temp `.part` identity, contiguous resume index, compact received ranges,
/// session binding, and expiry under the app support directory. Process restart
/// can reload valid partial receives; `invalidatePeer` clears checkpoints on
/// trust revoke/reset.
public final class TransferCheckpointStore {

    public enum Direction: String {
        case inbound
        case outbound
    }

    public struct Record: Equatable {
        public let fileId: String
        public let manifest: TransferManifest
        public let peerFingerprintHex: String
        public let sessionId: String
        public let nextChunkIndex: Int
        public let receivedRanges: [Range<Int>]
        public let updatedAtMs: Int64
        public let expiresAtMs: Int64
        public let direction: Direction
        public let outboundSourcePath: String?

        public init(
            fileId: String,
            manifest: TransferManifest,
            peerFingerprintHex: String,
            sessionId: String,
            nextChunkIndex: Int,
            receivedRanges: [Range<Int>],
            updatedAtMs: Int64,
            expiresAtMs: Int64,
            direction: Direction = .inbound,
            outboundSourcePath: String? = nil
        ) {
            self.fileId = fileId
            self.manifest = manifest
            self.peerFingerprintHex = peerFingerprintHex
            self.sessionId = sessionId
            self.nextChunkIndex = nextChunkIndex
            self.receivedRanges = receivedRanges
            self.updatedAtMs = updatedAtMs
            self.expiresAtMs = expiresAtMs
            self.direction = direction
            self.outboundSourcePath = outboundSourcePath
        }
    }

    public static let defaultTTLMs: Int64 = 7 * 24 * 60 * 60 * 1000

    private let root: URL
    private let ttlMs: Int64
    private let nowMs: () -> Int64
    private let lock = NSLock()

    public init(
        root: URL = TransferCheckpointStore.defaultRoot(),
        ttlMs: Int64 = TransferCheckpointStore.defaultTTLMs,
        nowMs: @escaping () -> Int64 = { Int64(Date().timeIntervalSince1970 * 1000) }
    ) {
        self.root = root
        self.ttlMs = ttlMs
        self.nowMs = nowMs
        try? FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
    }

    public static func defaultRoot() -> URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? FileManager.default.temporaryDirectory
        return base.appendingPathComponent("Hyphen/transfer-checkpoints", isDirectory: true)
    }

    public func save(_ record: Record) throws {
        lock.lock()
        defer { lock.unlock() }
        let data = try JSONSerialization.data(withJSONObject: encode(record), options: [.sortedKeys])
        try data.write(to: fileURL(for: record.fileId), options: .atomic)
    }

    public func load(fileId: String) -> Record? {
        lock.lock()
        defer { lock.unlock() }
        guard let data = try? Data(contentsOf: fileURL(for: fileId)),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let record = try? decode(object) else {
            return nil
        }
        return isExpired(record) ? nil : record
    }

    public func loadActive(for peerFingerprint: Data, direction: Direction? = nil) -> [Record] {
        lock.lock()
        defer { lock.unlock() }
        let hex = peerFingerprint.hexLower()
        guard let files = try? FileManager.default.contentsOfDirectory(
            at: root,
            includingPropertiesForKeys: nil
        ) else {
            return []
        }
        return files.compactMap { url -> Record? in
            guard url.pathExtension == "json" else { return nil }
            guard let data = try? Data(contentsOf: url),
                  let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let record = try? decode(object) else {
                return nil
            }
            guard record.peerFingerprintHex == hex else { return nil }
            if let direction, record.direction != direction { return nil }
            return isExpired(record) ? nil : record
        }
    }

    public func delete(fileId: String) {
        lock.lock()
        defer { lock.unlock() }
        try? FileManager.default.removeItem(at: fileURL(for: fileId))
    }

    public func invalidatePeer(_ peerFingerprint: Data) {
        lock.lock()
        defer { lock.unlock() }
        let hex = peerFingerprint.hexLower()
        guard let files = try? FileManager.default.contentsOfDirectory(
            at: root,
            includingPropertiesForKeys: nil
        ) else {
            return
        }
        for url in files where url.pathExtension == "json" {
            guard let data = try? Data(contentsOf: url),
                  let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let record = try? decode(object),
                  record.peerFingerprintHex == hex else {
                continue
            }
            try? FileManager.default.removeItem(at: url)
        }
    }

    public func invalidateAll() {
        lock.lock()
        defer { lock.unlock() }
        guard let files = try? FileManager.default.contentsOfDirectory(
            at: root,
            includingPropertiesForKeys: nil
        ) else {
            return
        }
        for url in files where url.pathExtension == "json" {
            try? FileManager.default.removeItem(at: url)
        }
    }

    public func purgeExpired() {
        lock.lock()
        defer { lock.unlock() }
        guard let files = try? FileManager.default.contentsOfDirectory(
            at: root,
            includingPropertiesForKeys: nil
        ) else {
            return
        }
        for url in files where url.pathExtension == "json" {
            if let data = try? Data(contentsOf: url),
               let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let record = try? decode(object),
               !isExpired(record) {
                continue
            }
            try? FileManager.default.removeItem(at: url)
        }
    }

    private func fileURL(for fileId: String) -> URL {
        root.appendingPathComponent("\(fileId).json")
    }

    private func isExpired(_ record: Record) -> Bool {
        nowMs() >= record.expiresAtMs
    }

    private func encode(_ record: Record) -> [String: Any] {
        var payload: [String: Any] = [
            "version": 1,
            "fileId": record.fileId,
            "manifest": record.manifest.payload,
            "peerFingerprintHex": record.peerFingerprintHex,
            "sessionId": record.sessionId,
            "nextChunkIndex": record.nextChunkIndex,
            "receivedRanges": record.receivedRanges.map { [$0.lowerBound, $0.upperBound] },
            "updatedAtMs": record.updatedAtMs,
            "expiresAtMs": record.expiresAtMs,
            "direction": record.direction.rawValue,
        ]
        if let outboundSourcePath = record.outboundSourcePath {
            payload["outboundSourcePath"] = outboundSourcePath
        }
        return payload
    }

    private func decode(_ object: [String: Any]) throws -> Record {
        guard try checkpointInt(object, "version") == 1 else {
            throw TransferError.invalidPayload("unsupported checkpoint version")
        }
        let directionRaw = try checkpointString(object, "direction")
        let direction = Direction(rawValue: directionRaw) ?? .inbound
        let rangeArrays = object["receivedRanges"] as? [[Any]] ?? []
        let ranges = try rangeArrays.map { pair -> Range<Int> in
            guard pair.count == 2 else {
                throw TransferError.invalidPayload("invalid received range")
            }
            let start = try intValue(pair[0])
            let endExclusive = try intValue(pair[1])
            guard endExclusive > start else {
                throw TransferError.invalidPayload("invalid received range bounds")
            }
            return start..<endExclusive
        }
        let outboundSourcePath = object["outboundSourcePath"] as? String
        return Record(
            fileId: try checkpointString(object, "fileId"),
            manifest: try TransferManifest(payload: object["manifest"] as? [String: Any] ?? [:]),
            peerFingerprintHex: try checkpointString(object, "peerFingerprintHex"),
            sessionId: try checkpointString(object, "sessionId"),
            nextChunkIndex: try checkpointInt(object, "nextChunkIndex"),
            receivedRanges: ranges,
            updatedAtMs: try checkpointInt64(object, "updatedAtMs"),
            expiresAtMs: try checkpointInt64(object, "expiresAtMs"),
            direction: direction,
            outboundSourcePath: outboundSourcePath
        )
    }
}

private func checkpointString(_ payload: [String: Any], _ field: String) throws -> String {
    guard let value = payload[field] as? String else {
        throw TransferError.invalidPayload("\(field) must be string")
    }
    return value
}

private func checkpointInt(_ payload: [String: Any], _ field: String) throws -> Int {
    Int(try checkpointInt64(payload, field))
}

private func checkpointInt64(_ payload: [String: Any], _ field: String) throws -> Int64 {
    guard let value = payload[field] else {
        throw TransferError.invalidPayload("\(field) must be integer")
    }
    if let int = value as? Int { return Int64(int) }
    if let int64 = value as? Int64 { return int64 }
    if let number = value as? NSNumber { return number.int64Value }
    throw TransferError.invalidPayload("\(field) must be integer")
}

private extension Data {
    func hexLower() -> String {
        map { String(format: "%02x", $0) }.joined()
    }
}

private func intValue(_ value: Any) throws -> Int {
    if let int = value as? Int { return int }
    if let int64 = value as? Int64 { return Int(int64) }
    if let num = value as? NSNumber { return num.intValue }
    throw TransferError.invalidPayload("invalid integer")
}

File: apps/android/app/src/main/kotlin/dev/hyphen/android/transfer/TransferCheckpointStore.kt
package dev.hyphen.android.transfer

import dev.hyphen.android.transport.Json
import java.io.File

/**
 * Peer-bound durable transfer checkpoints (dimension 02 P2). Stores manifest,
 * temp `.part` identity, contiguous resume index, compact received ranges,
 * session binding, and expiry under app-private storage. Process restart can
 * reload valid partial receives; [invalidatePeer] clears checkpoints on trust
 * revoke/reset.
 */
class TransferCheckpointStore(
    private val root: File,
    private val nowMs: () -> Long = System::currentTimeMillis,
    private val ttlMs: Long = DEFAULT_TTL_MS,
) {
    data class Record(
        val fileId: String,
        val manifest: TransferManifest,
        val peerFingerprintHex: String,
        val sessionId: String,
        val nextChunkIndex: Int,
        val receivedRanges: List<IntRange>,
        val updatedAtMs: Long,
        val expiresAtMs: Long,
        val direction: Direction = Direction.INBOUND,
        val outboundSourcePath: String? = null,
    ) {
        enum class Direction { INBOUND, OUTBOUND }
    }

    init {
        require(root.exists() || root.mkdirs()) { "transfer checkpoint storage unavailable" }
    }

    @Synchronized
    fun save(record: Record) {
        File(root, fileName(record.fileId)).writeText(encode(record))
    }

    @Synchronized
    fun load(fileId: String): Record? {
        val file = File(root, fileName(fileId))
        if (!file.isFile) return null
        return runCatching { decode(file.readText()) }
            .getOrNull()
            ?.takeUnless { isExpired(it) }
    }

    @Synchronized
    fun loadActiveForPeer(peerFingerprint: ByteArray, direction: Record.Direction? = null): List<Record> {
        val hex = peerFingerprint.toHexLower()
        return root.listFiles()
            .orEmpty()
            .asSequence()
            .filter { it.isFile && it.name.endsWith(".json") }
            .mapNotNull { runCatching { decode(it.readText()) }.getOrNull() }
            .filter { it.peerFingerprintHex == hex }
            .filter { direction == null || it.direction == direction }
            .filterNot { isExpired(it) }
            .toList()
    }

    @Synchronized
    fun delete(fileId: String) {
        File(root, fileName(fileId)).delete()
    }

    @Synchronized
    fun invalidatePeer(peerFingerprint: ByteArray) {
        val hex = peerFingerprint.toHexLower()
        root.listFiles()?.forEach { file ->
            if (!file.isFile || !file.name.endsWith(".json")) return@forEach
            val record = runCatching { decode(file.readText()) }.getOrNull() ?: return@forEach
            if (record.peerFingerprintHex == hex) {
                file.delete()
            }
        }
    }

    @Synchronized
    fun purgeExpired() {
        root.listFiles()?.forEach { file ->
            if (!file.isFile || !file.name.endsWith(".json")) return@forEach
            val record = runCatching { decode(file.readText()) }.getOrNull()
            if (record == null || isExpired(record)) {
                file.delete()
            }
        }
    }

    private fun isExpired(record: Record): Boolean = nowMs() >= record.expiresAtMs

    private fun fileName(fileId: String): String = "$fileId.json"

    private fun encode(record: Record): String {
        val ranges = Json.Arr(
            record.receivedRanges.map { range ->
                Json.Arr(listOf(Json.Num(range.first.toString()), Json.Num((range.last + 1).toString())))
            },
        )
        val fields = linkedMapOf(
            "version" to Json.Num("1"),
            "fileId" to Json.Str(record.fileId),
            "manifest" to record.manifest.toJson(),
            "peerFingerprintHex" to Json.Str(record.peerFingerprintHex),
            "sessionId" to Json.Str(record.sessionId),
            "nextChunkIndex" to Json.Num(record.nextChunkIndex.toString()),
            "receivedRanges" to ranges,
            "updatedAtMs" to Json.Num(record.updatedAtMs.toString()),
            "expiresAtMs" to Json.Num(record.expiresAtMs.toString()),
            "direction" to Json.Str(record.direction.name.lowercase()),
        )
        if (record.outboundSourcePath != null) {
            fields["outboundSourcePath"] = Json.Str(record.outboundSourcePath)
        }
        return Json.Obj(fields).encode()
    }

    private fun decode(text: String): Record {
        val obj = Json.parse(text) as Json.Obj
        require(int(obj, "version") == 1) { "unsupported checkpoint version" }
        val direction = when (string(obj, "direction")) {
            "outbound" -> Record.Direction.OUTBOUND
            else -> Record.Direction.INBOUND
        }
        val ranges = (obj["receivedRanges"] as Json.Arr).items.map { item ->
            val pair = item as Json.Arr
            val start = int(pair.items[0] as Json.Num)
            val endExclusive = int(pair.items[1] as Json.Num)
            require(endExclusive > start) { "invalid received range" }
            start until endExclusive
        }
        return Record(
            fileId = string(obj, "fileId"),
            manifest = TransferManifest.fromJson(obj["manifest"] as Json.Obj),
            peerFingerprintHex = string(obj, "peerFingerprintHex"),
            sessionId = string(obj, "sessionId"),
            nextChunkIndex = int(obj, "nextChunkIndex"),
            receivedRanges = ranges,
            updatedAtMs = long(obj, "updatedAtMs"),
            expiresAtMs = long(obj, "expiresAtMs"),
            direction = direction,
            outboundSourcePath = (obj["outboundSourcePath"] as? Json.Str)?.value,
        )
    }

    companion object {
        const val DEFAULT_TTL_MS = 7L * 24 * 60 * 60 * 1000

        fun defaultRoot(context: android.content.Context): File =
            File(context.filesDir, "transfer-checkpoints")
    }
}

private fun ByteArray.toHexLower(): String = joinToString("") { "%02x".format(it) }

private fun string(payload: Json.Obj, field: String): String =
    (payload[field] as? Json.Str)?.value ?: throw IllegalArgumentException("$field must be string")

private fun long(payload: Json.Obj, field: String): Long =
    (payload[field] as? Json.Num)?.asLong() ?: throw IllegalArgumentException("$field must be integer")

private fun int(payload: Json.Obj, field: String): Int =
    long(payload, field).also {
        require(it in Int.MIN_VALUE..Int.MAX_VALUE) { "$field out of int range" }
    }.toInt()

private fun int(num: Json.Num): Int =
    num.asLong()?.toInt() ?: throw IllegalArgumentException("invalid integer")
